File: 프로그래머스/lv2/최원준/개인정보 수집 유효기간.py
# https://www.acmicpc.net/problem/

'''
1. 아이디어 :

2. 시간복잡도 :
    O(
3. 자료구조 :

'''


def solution(today, terms, privacies):
    term_dict = {}
    for term in terms:
        key, val = term.split(" ")
        term_dict[key] = int(val) * 28

    def serialize(ymd):
        y,m,d = ymd.split(".")
        return int(y)*12*28 + int(m)*28 + int(d)

    today = serialize(today)

    ans=[]
    for i, privacy in enumerate(privacies):
        date, term = privacy.split(" ")
        if serialize(date) + term_dict[term] <= today:
            ans.append(i+1)
    return ans

File: 프로그래머스/lv1/이하은/개인정보 수집 유효기간.kt
class Solution {
    fun calcTotalDays(date: List<Int>): Int {
        return date[0] * 12 * 28 + date[1] * 28 + date[2]
    }
    
    fun solution(today: String, terms: Array<String>, privacies: Array<String>): List<Int> {
        val answer = mutableListOf<Int>()
        
        val todayDays = calcTotalDays(
            today.split(".").map{ it.toInt() }
        )
        
        val hash = hashMapOf<String, Int>() 
        for(term in terms){
            val temp = term.split(" ")
            hash[temp[0]] = temp[1].toInt()
        }
        
        privacies.forEachIndexed { idx, privacy -> 
            val temp = privacy.split(" ")
            val collectDays = calcTotalDays(
                temp[0].split(".").map{ it.toInt() }
            )
            val termType = temp[1]
            
            // 수집 일자 + 약관의 유효 기간
            val month = hash[termType]!!
            val expireDays = collectDays + (month * 28) - 1

            // 파기할 개인정보 번호 저장
            if(todayDays > expireDays){
                answer.add(idx + 1)
            }
        }
        
        return answer
    }
}
